beaconChain/lib/stateTransition/epoch/helpers.ts

```diff
@@ -0,0 +1,157 @@
+  // The total effective balance of active validators
+  const totalBalance = getTotalBalance(state, activeValidatorIndices);
+
+  // The maximum balance chrun in Gwei (for deposits and exists separately)
+  const a = new BN(MAX_DEPOSIT_AMOUNT);
+  const b = totalBalance.divn(2 * MAX_BALANCE_CHURN_QUOTIENT);
+  const maxBalanceChrun = a.gt(b) ? a : b;
+
+  // Activate validators within the allowable balance churn
+  let balanceChurn = new BN(0);
+  state.validatorRegistry.forEach((validator, index) => {
+    if (validator.activationEpoch > getEntryExitEffectEpoch(currentEpoch) && state.validatorBalances[index].lten(MAX_DEPOSIT_AMOUNT)) {
+      // Check the balance churn would be within the allowance
+      balanceChurn = balanceChurn.add(getEffectiveBalance(state, new BN(index)));
+      if (balanceChurn.gt(maxBalanceChrun)) {
+        return;
+      }
+      // Activate Validator
+      activateValidator(state, new BN(index), false);
+    }
+  });
+
+  // Exit validators within the allowable balance churn
+  balanceChurn = new BN(0);
+  state.validatorRegistry.forEach((validator: Validator, i) => {
+    const index: ValidatorIndex = new BN(i);
+    if (validator.exitEpoch > getEntryExitEffectEpoch(currentEpoch) && validator.statusFlags.and(INITIATED_EXIT)) {
+      // Check the balance churn would be within the allowance
+      balanceChurn = balanceChurn.add(getEffectiveBalance(state, index));
+      if (balanceChurn.gt(maxBalanceChrun)) {
+        return;
+      }
+      // Exit Validator
+      exitValidator(state, index);
+    }
+  });
+
+  state.validatorRegistryUpdateEpoch = currentEpoch;
+}
+
+/**
+ * Iterate through the validator registry and eject active validators with balance below EJECTION_BALANCE.
+ * @param {BeaconState} state
+ */
+export function processEjections(state: BeaconState): void {
+  for (let index of getActiveValidatorIndices(state.validatorRegistry, getCurrentEpoch(state))) {
+    if (state.validatorBalances[index.toNumber()].ltn(EJECTION_BALANCE)) {
+      exitValidator(state, index);
+    }
+  }
+}
+
+/**
+ * Returns the attestation with the lowest inclusion slot for a specified validatorIndex.
+ * @param {BeaconState} state
+ * @param {ValidatorIndex} validatorIndex
+ * @returns {uint64}
+ */
+export function inclusionSlot(state: BeaconState, validatorIndex: ValidatorIndex): uint64 {
+  let lowestInclusionSlot: uint64;
+  state.latestAttestations.forEach((attestation: PendingAttestation) => {
+    getAttestationParticipants(state, attestation.data, attestation.aggregationBitfield)
+      .forEach((index: ValidatorIndex) => {
+        if (index.eq(validatorIndex) && attestation.inclusionSlot.lt(lowestInclusionSlot)) {
+          lowestInclusionSlot = attestation.inclusionSlot;
+        }
+      })
+  });
+  return lowestInclusionSlot;
+}
+
+export function inclusionDistance(state: BeaconState, validatorIndex: ValidatorIndex): uint64 {
```

Comment:
use inclusionSlot function  in here

Reply:
I thought so too, but incluson slot returns a validator index, and we need access to the actual validator in order to perform the comparison on the `return` at the end.

Thoughts?

Reply:
Oh I see, we could either split out the core functionality into a sub helper function (that returns an attestation) (maybe called lowestInclusionSlotAttestation),
have inclusionSlot call that function and return the piece it needs,
have inclusionDistance call that function and return the piece it needs.

Or at least add a comment to inclusionDistance mentioning that the code is basically copy-pasted from inclusionSlot.

Reply:
I'll fix everything else and come back to this